### PR TITLE
[ONNX] Passes to reuse existing fake mode if possible

### LINKED ISSUE #105467: [ONNX] Decomp Pass behaves differently w/ '_dynamo.export(aten_graph=True)' under dynamic shapes and pre-defined fake mode
https://github.com/pytorch/pytorch/pull/105246 requires `torch._dynamo.export(aten_graph=True)`. The reason is an issue with dynamic shape and fake tensor mode, as shown below.

The error below was first reported at https://github.com/pytorch/pytorch/issues/99774, but never really addressed. Instead, we adopted a workaround that now is failing with fake tensor.

The fix is probably "dynamizing" the input before calling the exporter, which might be done on a separate task

```bash
# pytest -sv test/onnx/test_fx_to_onnx_with_onnxruntime.py::TestFxToOnnxWithOnnxRuntime_op_level_debug_False_dynamic_shapes_False::test_large_scale_exporter_with_tiny_gpt2

____________________________ TestFxToOnnxWithOnnxRuntime_op_level_debug_False_dynamic_shapes_False.test_large_scale_exporter_with_tiny_gpt2 _____________________________
Traceback (most recent call last):
  File "/opt/pytorch/torch/onnx/_internal/diagnostics/infra/decorator.py", line 127, in wrapper
    return_values = fn(*args, **kwargs)
  File "/opt/pytorch/torch/onnx/_internal/fx/_pass.py", line 192, in run
    module = self._run(*args, **kwargs)
  File "<@beartype(torch.onnx._internal.fx.passes.decomp.Decompose._run) at 0x7ff60cd8cdc0>", line 11, in _run
  File "/opt/pytorch/torch/onnx/_internal/fx/passes/decomp.py", line 54, in _run
    decomposed_module = proxy_tensor.make_fx(
  File "/opt/pytorch/torch/fx/experimental/proxy_tensor.py", line 809, in wrapped
    t = dispatch_trace(wrap_key(func, args, fx_tracer, pre_dispatch), tracer=fx_tracer, concrete_args=tuple(phs))
  File "/opt/pytorch/torch/_compile.py", line 24, in inner
    return torch._dynamo.disable(fn, recursive)(*args, **kwargs)
  File "/opt/pytorch/torch/_dynamo/eval_frame.py", line 294, in _fn
    return fn(*args, **kwargs)
  File "/opt/pytorch/torch/_dynamo/external_utils.py", line 17, in inner
    return fn(*args, **kwargs)
  File "/opt/pytorch/torch/fx/experimental/proxy_tensor.py", line 468, in dispatch_trace
    graph = tracer.trace(root, concrete_args)
  File "/opt/pytorch/torch/_dynamo/eval_frame.py", line 294, in _fn
    return fn(*args, **kwargs)
  File "/opt/pytorch/torch/_dynamo/external_utils.py", line 17, in inner
    return fn(*args, **kwargs)
  File "/opt/pytorch/torch/fx/_symbolic_trace.py", line 817, in trace
    (self.create_arg(fn(*args)),),
  File "/opt/pytorch/torch/fx/experimental/proxy_tensor.py", line 485, in wrapped
    out = f(*tensors)
  File "<string>", line 1, in <lambda>
  File "/opt/pytorch/torch/onnx/_internal/fx/passes/_utils.py", line 30, in wrapped
    return torch.fx.Interpreter(graph_module).run(*args)
  File "/opt/pytorch/torch/fx/interpreter.py", line 138, in run
    self.env[node] = self.run_node(node)
  File "/opt/pytorch/torch/fx/interpreter.py", line 195, in run_node
    return getattr(self, n.op)(n.target, args, kwargs)
  File "/opt/pytorch/torch/fx/interpreter.py", line 267, in call_function
    return target(*args, **kwargs)
RuntimeError: Cannot call sizes() on tensor with symbolic sizes/strides

While executing %matmul : [num_users=1] = call_function[target=torch.matmul](args = (%permute, %transpose), kwargs = {})
Original traceback:
  File "/opt/conda/envs/ptca/lib/python3.8/site-packages/transformers/models/gpt2/modeling_gpt2.py", line 903, in forward
    outputs = block(
  File "/opt/pytorch/torch/nn/modules/module.py", line 1531, in _call_impl
    return forward_call(*args, **kwargs)
  File "/opt/conda/envs/ptca/lib/python3.8/site-packages/transformers/models/gpt2/modeling_gpt2.py", line 391, in forward
    attn_outputs = self.attn(
  File "/opt/pytorch/torch/nn/modules/module.py", line 1531, in _call_impl
    return forward_call(*args, **kwargs)
  File "/opt/conda/envs/ptca/lib/python3.8/site-packages/transformers/models/gpt2/modeling_gpt2.py", line 332, in forward
    attn_output, attn_weights = self._attn(query, key, value, attention_mask, head_mask)
  File "/opt/conda/envs/ptca/lib/python3.8/site-packages/transformers/models/gpt2/modeling_gpt2.py", line 184, in _attn
    attn_weights = torch.matmul(query, key.transpose(-1, -2))


The above exception was the direct cause of the following exception:

Traceback (most recent call last):
  File "/opt/pytorch/torch/onnx/_internal/exporter.py", line 831, in dynamo_export
    return Exporter(
  File "/opt/pytorch/torch/onnx/_internal/exporter.py", line 649, in export
    graph_module = pre_export_passes(
  File "<@beartype(torch.onnx._internal.exporter.pre_export_passes) at 0x7ff6a5c73d30>", line 97, in pre_export_passes
  File "/opt/pytorch/torch/onnx/_internal/exporter.py", line 864, in pre_export_passes
    module = passes.Decompose(
  File "/opt/pytorch/torch/onnx/_internal/diagnostics/infra/decorator.py", line 142, in wrapper
    ctx.log_and_raise_if_error(diag)
  File "/opt/pytorch/torch/onnx/_internal/diagnostics/infra/context.py", line 265, in log_and_raise_if_error
    raise RuntimeErrorWithDiagnostic(
torch.onnx._internal.diagnostics.infra.context.RuntimeErrorWithDiagnostic: Running Decompose pass. Raised from:
    RuntimeError: Cannot call sizes() on tensor with symbolic sizes/strides

While executing %matmul : [num_users=1] = call_function[target=torch.matmul](args = (%permute, %transpose), kwargs = {})
Original traceback:
  File "/opt/conda/envs/ptca/lib/python3.8/site-packages/transformers/models/gpt2/modeling_gpt2.py", line 903, in forward
    outputs = block(
  File "/opt/pytorch/torch/nn/modules/module.py", line 1531, in _call_impl
    return forward_call(*args, **kwargs)
  File "/opt/conda/envs/ptca/lib/python3.8/site-packages/transformers/models/gpt2/modeling_gpt2.py", line 391, in forward
    attn_outputs = self.attn(
  File "/opt/pytorch/torch/nn/modules/module.py", line 1531, in _call_impl
    return forward_call(*args, **kwargs)
  File "/opt/conda/envs/ptca/lib/python3.8/site-packages/transformers/models/gpt2/modeling_gpt2.py", line 332, in forward
    attn_output, attn_weights = self._attn(query, key, value, attention_mask, head_mask)
  File "/opt/conda/envs/ptca/lib/python3.8/site-packages/transformers/models/gpt2/modeling_gpt2.py", line 184, in _attn
    attn_weights = torch.matmul(query, key.transpose(-1, -2))


The above exception was the direct cause of the following exception:

Traceback (most recent call last):
  File "/opt/conda/envs/ptca/lib/python3.8/unittest/case.py", line 60, in testPartExecutor
    yield
  File "/opt/conda/envs/ptca/lib/python3.8/unittest/case.py", line 676, in run
    self._callTestMethod(testMethod)
  File "/opt/conda/envs/ptca/lib/python3.8/unittest/case.py", line 633, in _callTestMethod
    method()
  File "/opt/pytorch/torch/testing/_internal/common_utils.py", line 2354, in wrapper
    method(*args, **kwargs)
  File "/opt/pytorch/test/onnx/pytorch_test_common.py", line 231, in wrapper
    return func(self, *args, **kwargs)
  File "/opt/pytorch/test/onnx/test_fx_to_onnx_with_onnxruntime.py", line 965, in test_large_scale_exporter_with_tiny_gpt2
    self._test_fake_tensor_mode_exporter(
  File "<@beartype(test_fx_to_onnx_with_onnxruntime.TestFxToOnnxWithOnnxRuntime._test_fake_tensor_mode_exporter) at 0x7ff6077ba040>", line 108, in _test_fake_tensor_mode_exporter
  File "/opt/pytorch/test/onnx/test_fx_to_onnx_with_onnxruntime.py", line 883, in _test_fake_tensor_mode_exporter
    export_output = torch.onnx.dynamo_export(
  File "<@beartype(torch.onnx._internal.exporter.dynamo_export) at 0x7ff6a5c73c10>", line 53, in dynamo_export
  File "/opt/pytorch/torch/onnx/_internal/exporter.py", line 846, in dynamo_export
    raise OnnxExporterError(
torch.onnx.OnnxExporterError: Failed to export the model to ONNX. Generating SARIF report at report_dynamo_export.sarif. Please report a bug on PyTorch Github: https://github.com/pytorch/pytorch/issues

To execute this test, run the following from the base repo dir:
     python test/onnx/test_fx_to_onnx_with_onnxruntime.py -k test_large_scale_exporter_with_tiny_gpt2

This message can be suppressed by setting PYTORCH_PRINT_REPRO_ON_FAILURE=0
======================================================================== short test summary info ========================================================================
FAILED [107.9249s] test/onnx/test_fx_to_onnx_with_onnxruntime.py::TestFxToOnnxWithOnnxRuntime_op_level_debug_False_dynamic_shapes_False::test_large_scale_exporter_with_tiny_gpt2 - torch.onnx.OnnxExporterError: Failed to export the model to ONNX. Generating SARIF report at report_dynamo_export.sarif. Please report a bug on PyTorch Github: http...
===================================================================== 1 failed in 114.03s (0:01:54) =====================================================================
```

### PR DESCRIPTION
**Pull request by BowenBao**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #105769
* __->__ #105764

Fixes #105467, namely the need of setting `aten_graph=True` in `_dynamo.export`
to make fake mode onnx exporter work. 

Previously, `make_fx` called by passes always create new fake mode. Hence it is 
missing out info from `shape_env` recorded during dynamo export. This PR tries
to check and fetch existing fake mode from graph node meta.
Also enable python dispatcher context when calling `make_fx`. This is done in
`_dynamo.export(aten_graph=True)` but was missing in our passes.

### CODE REVIEW

File: torch/onnx/_internal/fx/_pass.py [hunk 1]
```diff
@@ -163,6 +164,31 @@ def __init__(
         """
         self.module = module
         self.diagnostic_context = diagnostic_context
+        self.fake_mode = self._detect_fake_mode()
```
Comment (thiagocrepaldi) [inline torch/onnx/_internal/fx/_pass.py:167]: missing docstring from `Args:` above.

**Review by thiagocrepaldi — COMMENTED** [review]
Does this change do anything other than just remove `aten_graph=` from `torch._dynamo.export` call and add equivalent code to make up for it?

Any new model can be converted or did it fix the FX IR? In that case, should we add a unit test

The resulting code seems to have moved the implementation from the main flag into our code.

**Comment by BowenBao** [comment]
Right, there is no extra things. To match aten_graph behavior accurately. 

I think the fundamental reason of such fake mode context setup is to preserve the shape_env info populated during dynamo export.

File: torch/onnx/_internal/fx/_pass.py (hunk 1 above)
Comment (BowenBao) [reply torch/onnx/_internal/fx/_pass.py:167]: it's not an arg but I'll add it to instance attribute

File: torch/onnx/_internal/fx/_pass.py [hunk 2]
```diff
@@ -161,8 +168,33 @@ def __init__(
...
+        """Detect fake mode from the graph.
+
+        Scan through all nodes in graph and their meta['val'] to detect fake mode.
+        """
+        fake_tensors = []
+        for node in self.module.graph.nodes:
+            try:
+                fake_tensors.append(node.meta.get("val"))
+            except RuntimeError:
+                continue
```
Comment (justinchuby) [inline torch/onnx/_internal/fx/_pass.py:185]: Comment on when the error is expected?

File: torch/onnx/_internal/fx/passes/decomp.py [hunk 3]
```diff
@@ -47,17 +49,27 @@ def _run(self, *args, **kwargs) -> torch.fx.GraphModule:
         #  mul = sym_size_2 * sym_size_3;  sym_size_2 = sym_size_3 = None
         #  view: f32[3, 5, 20] = torch.ops.aten.view.default(x, [sym_size, sym_size_1, mul])
 
-        fx_mode = "symbolic" if self.enable_dynamic_axes else "fake"
+        fake_mode: Any = self.fake_mode
+        maybe_fake_args = self._maybe_fakefy_args(fake_mode, *args)
+        if fake_mode is not None:
+            # Using existing fake mode as context, signal `make_fx` that it does not need
+            # to create a new fake mode by passing tracing_mode as "real".
+            tracing_mode = "real"
```
[the remark this answers is not in the document]
Comment (titaiwangms) [reply torch/onnx/_internal/fx/passes/decomp.py:57]: By the way, I think this if/else branch saves us from Fake_mode making dynamic_shapes=False meaningless.

File: torch/onnx/_internal/fx/passes/functionalization.py [hunk 4]
```diff
@@ -102,15 +103,26 @@ def _run(self, *args) -> torch.fx.GraphModule:
...
+        fake_mode: Any = self.fake_mode
+        maybe_fake_args = self._maybe_fakefy_args(fake_mode, *args)
+        if fake_mode is not None:
+            # Using existing fake mode as context, signal `make_fx` that it does not need
+            # to create a new fake mode by passing tracing_mode as "real".
+            tracing_mode = "real"
+        else:
+            # Existing fake mode not found, signal `make_fx` to create one.
+            fake_mode = contextlib.nullcontext()
+            tracing_mode = "symbolic" if self.enable_dynamic_axes else "fake"
```
Comment (titaiwangms) [inline torch/onnx/_internal/fx/passes/functionalization.py:116]: It's possible to make this to _pass.py? So we don't write it in passes?

**Review by titaiwangms — APPROVED** [review]

File: torch/onnx/_internal/fx/passes/decomp.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (BowenBao) [reply torch/onnx/_internal/fx/passes/decomp.py:57]: `fake` and `symbolic` both signals `make_fx` to create a new fake mode and use it as context. `real` does nothing.

> By the way, I think this if/else branch saves us from Fake_mode making dynamic_shapes=False meaningless.

Hmm, would it? I think `fake_mode` will not be none for fake export scenario, so it never goes to the branch of selecting `tracing_mode='fake'`.

File: torch/onnx/_internal/fx/passes/functionalization.py [hunk 5]
```diff
@@ -102,15 +103,26 @@ def _run(self, *args) -> torch.fx.GraphModule:
...
-
-        graph_module = proxy_tensor.make_fx(
-            functionalized_callable,
-            decomposition_table={},
-            tracing_mode=fx_mode,
-            _allow_non_fake_inputs=True,
-            _allow_fake_constant=self.allow_fake_constant,
-        )(*args)
+
+        fake_mode: Any = self.fake_mode
```
[the remark this answers is not in the document]
Comment (BowenBao) [reply torch/onnx/_internal/fx/passes/functionalization.py:107]: This is for mypy, because later it is overloaded with `contextlib.nullcontext`. I don't want to annotate with ridiculously long type annotations.

File: torch/onnx/_internal/fx/passes/functionalization.py (hunk 4 above)
Comment (BowenBao) [reply torch/onnx/_internal/fx/passes/functionalization.py:116]: true there is a bit of repetition. I'm slightly leaning towards leaving as is for now. In case there might be more changes to it.

File: torch/onnx/_internal/fx/_pass.py (hunk 2 above)
Comment (BowenBao) [reply torch/onnx/_internal/fx/_pass.py:185]: good catch, I think it was left over when the internal code was calling another function. I'll remove

File: torch/onnx/_internal/fx/passes/decomp.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (titaiwangms) [reply torch/onnx/_internal/fx/passes/decomp.py:57]: I see

Comment (thiagocrepaldi) [inline torch/onnx/_internal/fx/passes/decomp.py:57]: This also makes `make_fx` create a graph with static shape.
https://github.com/pytorch/pytorch/issues/105464 has a task to track fake mode + dynamic shape, but a TODO would make finding this integration point easier in the future

File: torch/onnx/_internal/fx/passes/functionalization.py (hunk 5 above)
[the remark this answers is not in the document]
Comment (thiagocrepaldi) [reply torch/onnx/_internal/fx/passes/functionalization.py:107]: `Any` is hacky because it deceives the type check system Consider `# type: ignore[xxx]` to make it clear that you want to bypass things

File: torch/onnx/_internal/fx/passes/functionalization.py [hunk 6]
```diff
@@ -102,15 +103,26 @@ def _run(self, *args) -> torch.fx.GraphModule:
...
+        maybe_fake_args = self._maybe_fakefy_args(fake_mode, *args)
+        if fake_mode is not None:
+            # Using existing fake mode as context, signal `make_fx` that it does not need
+            # to create a new fake mode by passing tracing_mode as "real".
+            tracing_mode = "real"
```
Comment (thiagocrepaldi) [inline torch/onnx/_internal/fx/passes/functionalization.py:112]: https://github.com/pytorch/pytorch/issues/105464 tracks fake mode + dynamic shape support which might change how `tracing_mode` and `fake_mode` interacts in near future.

I suggest a TODO to help finding this spot in the future when I revisit dyn shape+fake

File: torch/onnx/_internal/fx/passes/type_promotion.py [hunk 7]
```diff
@@ -1661,12 +1648,10 @@ def _run(self, *args, **kwargs) -> torch.fx.GraphModule:
         )
         assert not kwargs, "`kwargs` is not supported"
 
-        fake_tensor_mode = self._detect_fake_mode() or fake_tensor.FakeTensorMode(
-            allow_non_fake_inputs=True
-        )
         fake_args = self._fetch_fake_args()
+        fake_mode = self.fake_mode or torch._dynamo.utils.detect_fake_mode(fake_args)
```
Comment (thiagocrepaldi) [inline torch/onnx/_internal/fx/passes/type_promotion.py:1652]: it is a bit tricky knowing when we have to use `torch._dynamo.utils.detect_fake_mode` or `torch.onnx._internal.fx._pass._detect_fake_mode`

Maybe we should always use one of them in the onnx exporter to make it uniform? or name our version differently to make clear the distinction, such as `_detect_fake_mode_from_fx_module`

File: torch/onnx/_internal/fx/_pass.py [hunk 8]
```diff
@@ -161,8 +168,28 @@ def __init__(
...
+    def _maybe_fakefy_args(
+        self, fake_mode: Optional[fake_tensor.FakeTensorMode], *args: Any
+    ) -> Tuple[Any, ...]:
+        if fake_mode is None:
+            return args
+        # NB: This should hit the cache if tensors were fakefied before.
+        # E.g., when the fx graph is produced by Dynamo.
+        return tuple(
+            fake_mode.from_tensor(t) if isinstance(t, torch.Tensor) else t for t in args
+        )
```
Comment (thiagocrepaldi) [inline torch/onnx/_internal/fx/_pass.py:192]: Maybe move this to a `fake_tensor.py` or `fake_util.py` file? This is a domain specific util not really related to passes only. It is hard to reuse a generic util that is hidden into an unrelated file

This is used by `_pass.py` and maybe (see comment below) should be used by `type_promotion.py`; it might be used at another components when dynamic shape is fully supported

```suggestion
    def _detect_fake_mode(fx_module) -> Optional[fake_tensor.FakeTensorMode]:
        """Detect fake mode from the graph.

        Scan through all nodes in graph and their meta['val'] to detect fake mode.
        """
        fake_tensors = [node.meta.get("val") for node in self.module.graph.nodes]
        return torch._dynamo.utils.detect_fake_mode(fake_tensors)

    def _maybe_fakefy_args(
        fake_mode: Optional[fake_tensor.FakeTensorMode], *args: Any
    ) -> Tuple[Any, ...]:
        if fake_mode is None:
            return args
        # NB: This should hit the cache if tensors were fakefied before.
        # E.g., when the fx graph is produced by Dynamo.
        return tuple(
            fake_mode.from_tensor(t) if isinstance(t, torch.Tensor) else t for t in args
        )
```

Comment (BowenBao) [reply torch/onnx/_internal/fx/_pass.py:192]: Similar to reply at https://github.com/pytorch/pytorch/pull/105764#discussion_r1271612064, but I will do this in a follow up.

File: torch/onnx/_internal/fx/passes/type_promotion.py (hunk 7 above)
Comment (BowenBao) [reply torch/onnx/_internal/fx/passes/type_promotion.py:1652]: This check with dynamo utils is redundant. I'll remove.

File: torch/onnx/_internal/fx/passes/decomp.py (hunk 3 above)
Comment (BowenBao) [reply torch/onnx/_internal/fx/passes/decomp.py:57]: This does not make static shape.

This allows reusing the dynamo administrative fake mode, which contains dynamic shape info if dynamic shape is turned on.

File: torch/onnx/_internal/fx/passes/functionalization.py (hunk 5 above)
[the remark this answers is not in the document]
Comment (BowenBao) [reply torch/onnx/_internal/fx/passes/functionalization.py:107]: Both deceives type check system. I felt this is simple enough code for a single Any. I will change to multiple lines of ignore if folks prefer.

File: torch/onnx/_internal/fx/passes/functionalization.py (hunk 6 above)
Comment (BowenBao) [reply torch/onnx/_internal/fx/passes/functionalization.py:112]: I can add one if it helps. But be sure to remove the TODO once we get more clarity on fake + dynamic.

File: torch/onnx/_internal/fx/passes/functionalization.py (hunk 5 above)
[the remark this answers is not in the document]
Comment (titaiwangms) [reply torch/onnx/_internal/fx/passes/functionalization.py:107]: I don't have strong preference. But even non-annotated better than Any.

File: torch/onnx/_internal/fx/passes/functionalization.py [hunk 9]
```diff
@@ -102,15 +104,29 @@ def _run(self, *args) -> torch.fx.GraphModule:
...
+        else:
+            # Existing fake mode not found, signal `make_fx` to create one.
+            fake_mode = contextlib.nullcontext()  # type: ignore[assignment]
+            tracing_mode = "symbolic" if self.enable_dynamic_axes else "fake"
+
+        assert fake_mode is not None
```
Comment (thiagocrepaldi) [inline torch/onnx/_internal/fx/passes/functionalization.py:121]:
```suggestion
        assert fake_mode is not None, "Cannot detect fake mode."
```

File: torch/onnx/_internal/fx/passes/decomp.py [hunk 10]
```diff
@@ -47,17 +50,30 @@ def _run(self, *args, **kwargs) -> torch.fx.GraphModule:
...
-        # Make sure the feed-in "module" is stateless.
-        decomposed_module = proxy_tensor.make_fx(
-            module,
-            decomposition_table=self.decomposition_table,
-            tracing_mode=fx_mode,
-            _allow_non_fake_inputs=True,
-            _allow_fake_constant=self.allow_fake_constant,
-        )(*args)
+        assert fake_mode is not None
+        with python_dispatch.enable_python_dispatcher(), fake_mode:
```
Comment (thiagocrepaldi) [inline torch/onnx/_internal/fx/passes/decomp.py:68]: out of curiosity, which scenario requires the python dispatcher to be explicitly enabled for the pass (and not for the actual call to `torch._dynamo.export`)?

My understanding it is already enabled by default, so this would be needed if the user explicitly disables it, right? In that scenario, should we re-enable it?

File: torch/onnx/_internal/fx/passes/functionalization.py (hunk 9 above)
Comment (BowenBao) [reply torch/onnx/_internal/fx/passes/functionalization.py:121]: This is impossible for fake_mode to be None, however mypy couldn't figure out.

Comment (BowenBao) [inline torch/onnx/_internal/fx/passes/functionalization.py:121]:
```suggestion
        assert fake_mode is not None  # for mypy
```

File: torch/onnx/_internal/fx/passes/decomp.py [hunk 11]
```diff
@@ -47,17 +50,30 @@ def _run(self, *args, **kwargs) -> torch.fx.GraphModule:
...
         # Apply decomposition table to the input graph.
-        # Make sure the feed-in "module" is stateless.
-        decomposed_module = proxy_tensor.make_fx(
-            module,
-            decomposition_table=self.decomposition_table,
-            tracing_mode=fx_mode,
-            _allow_non_fake_inputs=True,
-            _allow_fake_constant=self.allow_fake_constant,
-        )(*args)
+        assert fake_mode is not None
```
Comment (BowenBao) [inline torch/onnx/_internal/fx/passes/decomp.py:67]:
```suggestion
        assert fake_mode is not None  # for mypy
```

File: torch/onnx/_internal/fx/passes/decomp.py (hunk 10 above)
Comment (BowenBao) [reply torch/onnx/_internal/fx/passes/decomp.py:68]: Great question. It is disabled by default. Selectively turned on when one wishes to enable the python dispatcher key in the dispatcher key stack. From export user perspective this should be super low level detail and not exposed.

File: torch/onnx/_internal/fx/passes/functionalization.py (hunk 9 above)
Comment (BowenBao) [reply torch/onnx/_internal/fx/passes/functionalization.py]: Used github commit changes for this. I'm not sure if it gets picked up by ghstack and merge bot, let's find out.

Comment (BowenBao) [reply torch/onnx/_internal/fx/passes/functionalization.py]: ok that didn't work

File: torch/onnx/_internal/fx/passes/decomp.py (hunk 10 above)
Comment (thiagocrepaldi) [reply torch/onnx/_internal/fx/passes/decomp.py:68]: I didn't mean visible to the user, but something like

```python

class Exporter:

    def export():
        with python_dispatch.enable_python_dispatcher():
            pre_export_passes()
            torch._dynamo.export()
```

Comment (thiagocrepaldi) [reply torch/onnx/_internal/fx/passes/decomp.py:68]: That is, why we are internally enabling the dispatcher only for a few passes while not to the `fx_tracer.generate_fx()` itself. There is usually a "sync" between `make_fx` and `torch._dynamo.export` calls to keep their fake_modeness identical

Comment (BowenBao) [reply torch/onnx/_internal/fx/passes/decomp.py:68]: Short answer: this matches with aten_graph=True behavior.
Longer answer: `torch._dynamo.export` manages `enable_python_dispatcher` internally. `make_fx` on the other hand can be viewed as a lower level api. While It does have convenience argument `tracing_mode` that sets up fake mode and related configurations inside, one should construct and re-use existing contexts from `torch._dynamo.export` in order to achieve the synchronization regarding fake mode and friends.
